src/Symfony/Component/Security/Http/Firewall/AccessListener.php

```diff
@@ -24,23 +27,86 @@
  */
 class AccessListener implements ListenerInterface
 {
+    /**
+     * @var TokenStorageInterface
+     * @deprecated
+     */
+    private $tokenStorage;
+
+    /**
+     * @var AccessDecisionManagerInterface
+     * @deprecated
+     */
+    private $accessDecisionManager;
+
+    /**
+     * @var AuthenticationManagerInterface
+     * @deprecated
+     */
+    private $authManager;
+
+    /** @var AccessMapInterface */
     private $map;
+
+    /** @var AuthorizationCheckerInterface */
     private $authorizationChecker;
 
-    public function __construct(AuthorizationCheckerInterface $authorizationChecker, AccessMapInterface $map)
+    /**
+     * AccessListener constructor.
+     *
+     * @param TokenStorageInterface|AuthorizationCheckerInterface $tokenStorage
+     * @param AccessDecisionManagerInterface|AccessMapInterface   $accessDecisionManager
+     * @param AccessMapInterface|null                             $map
+     * @param AuthenticationManagerInterface|null                 $authManager
+     *
+     * @throws \BadMethodCallException
+     */
+    public function __construct($tokenStorage, $accessDecisionManager, $map = null, $authManager = null)
     {
-        $this->map = $map;
-        $this->authorizationChecker = $authorizationChecker;
+        $oldSignature = $tokenStorage instanceof TokenStorageInterface
+            && $accessDecisionManager instanceof AccessDecisionManagerInterface
+            && $map instanceof AccessMapInterface
+            && $authManager instanceof AuthenticationManagerInterface;
+
+        $newSignature = $tokenStorage instanceof AuthorizationCheckerInterface
+            && $accessDecisionManager instanceof AccessMapInterface
+            && null === $map
+            && null === $authManager;
+
+        // confirm possible signatures
+        if (!$oldSignature && !$newSignature) {
+            throw new \BadMethodCallException('Unable to construct AccessListener, please provide the correct arguments');
+        }
+
+        if ($oldSignature) {
+            @trigger_error('The '.__CLASS__.' constructor now expects 2 arguments', E_USER_DEPRECATED);
```

Comment:
This message should mentioned what's deprecated as of when and how to fix it

Reply:
Updated deprecation message, but I would like to get suggestions on better wording